File: 백준/Silver/9012. 괄호/괄호.py
import sys


def is_vps(string):
    stack = []
    for letter in string:
        if letter == '(':
            stack.append(letter)
        else:
            if not stack:
                return 'NO'
            else:
                stack.pop()
    if stack:
        return 'NO'
    else:
        return 'YES'


t = int(sys.stdin.readline())
for _ in range(t):
    string = sys.stdin.readline().rstrip()
    print(is_vps(string))

File: 백준/Silver/9012. 괄호/괄호.cc
#include <bits/stdc++.h>
using namespace std;

int N;
string s;

int main(){
    ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    cin >> N;
    while (N--){
        stack<char> stk;
        cin >> s;
        for (char ch : s) {
            if (ch == ')'){
                if (stk.size() && stk.top() == '(') stk.pop();
                else stk.push(ch);
            }
            else stk.push(ch);
        }
        if (stk.size()) cout << "NO" << '\n';
        else cout << "YES" << '\n';
    }
    return 0;
}